lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1117,45 +1106,63 @@ fn find_best_match<'a>(
 
 #[allow(clippy::too_many_arguments)]
 fn better_match(
-    my_pattern: &ColorPattern,
-    candidate_pattern: &ColorPattern,
-    my_guide_pattern: &ColorPattern,
-    candidate_guide_pattern: &ColorPattern,
+    k_neighs: &[(SignedCoord2D, MapId)],
+    outside_color: image::Rgba<u8>,
+    source_maps: &[ImageBuffer<'_>],
+    guides: &Option<GuidesStruct<'_>>,
+    my_precomputed_pattern: &ColorPattern,
+    my_precomputed_guide_pattern: &ColorPattern,
     distance_gaussians: &[f32], //weight by distance
     my_cost: &PrerenderedU8Function,
     guide_cost: Option<&PrerenderedU8Function>,
     current_best: f32,
 ) -> Option<f32> {
     let mut score: f32 = 0.0; //minimize score
 
-    for ((my_value, candidate_value), dist_gaussian) in my_pattern
-        .0
-        .iter()
-        .copied()
-        .zip(candidate_pattern.0.iter().copied())
-        .zip(distance_gaussians.iter().copied())
-    {
-        score += dist_gaussian * my_cost.get(my_value, candidate_value);
+    let mut i = 0;
+    let mut next_pixel = [0; 4];
+    let mut next_pixel_score: f32;
+    for (n_coord, n_map) in k_neighs {
+        next_pixel_score = 0.0;
+        let end = i + 4;
 
-        if score >= current_best {
-            return None;
+        //check if he haven't gone outside the possible bounds
+        get_color_of_neighbor(
+            outside_color,
+            source_maps,
+            *n_coord,
+            *n_map,
+            &mut next_pixel,
+            false,
+            (0, 0),
+        );
+
+        for (channel_n, &channel) in next_pixel.iter().enumerate() {
+            next_pixel_score += my_cost.get(my_precomputed_pattern.0[i + channel_n], channel);
         }
-    }
 
-    if let Some(guide_cost_fn) = guide_cost {
-        for ((my_guide, candidate_guide), dist_gaussian) in my_guide_pattern
-            .0
-            .iter()
-            .copied()
-            .zip(candidate_guide_pattern.0.iter().copied())
-            .zip(distance_gaussians.iter().copied())
-        {
-            score += dist_gaussian * guide_cost_fn.get(my_guide, candidate_guide);
+        if let Some(guide_cost) = guide_cost {
+            let example_guides = &(guides.as_ref().unwrap().example_guides);
+            get_color_of_neighbor(
+                outside_color,
+                example_guides,
+                *n_coord,
+                *n_map,
+                &mut next_pixel,
+                false,
+                (0, 0),
+            );
 
-            if score >= current_best {
-                return None;
+            for (channel_n, &channel) in next_pixel.iter().enumerate() {
+                next_pixel_score +=
+                    guide_cost.get(my_precomputed_guide_pattern.0[i + channel_n], channel);
             }
         }
+        score += next_pixel_score * distance_gaussians[i];
```

Comment:
should probably shrink distance gaussians from size (num neighbors) * 4 to just num neighbors as that is all I'm using here